### PR TITLE
[xray] Implement faster flush policy for lineage cache

### PR DESCRIPTION
**Pull request by stephanie-wang**
## What do these changes do?

Lineage is stored locally at each node in the lineage cache, but it must be flushed to a persistent store to prevent the caches from filling up. This implements a policy that flushes each task immediately to the persistent store, rather than flushing tasks in data dependency order.

## Related issue number
This is part 1 of #2948.

### CODE REVIEW

File: src/ray/raylet/lineage_cache.cc [hunk 1]
```diff
@@ -180,34 +177,62 @@ void MergeLineageHelper(const TaskID &task_id, const Lineage &lineage_from,
...
-      UnsubscribeTask(task_id);
-    }
+  // Merge the uncommitted lineage into the lineage cache. Collect the IDs of
+  // tasks that we should subscribe to. These are all of the tasks that were
+  // included in the uncommitted lineage that we did not already have in our
+  // stash.
+  std::unordered_set<TaskID> subscribe_tasks;
+  AddUncommittedLineage(task_id, uncommitted_lineage, subscribe_tasks);
+
+  // Do not subscribe to the waiting task.
```
Comment (richardliaw) [inline src/ray/raylet/lineage_cache.cc]: mention that the task is local so it doesn't make sense to subscribe to it?

File: src/ray/raylet/lineage_cache.cc [hunk 2]
```diff
@@ -139,10 +139,7 @@ LineageCache::LineageCache(const ClientID &client_id,
...
                            gcs::PubsubInterface<TaskID> &task_pubsub,
                            uint64_t max_lineage_size)
-    : client_id_(client_id),
-      task_storage_(task_storage),
-      task_pubsub_(task_pubsub),
-      max_lineage_size_(max_lineage_size) {}
+    : client_id_(client_id), task_storage_(task_storage), task_pubsub_(task_pubsub) {}
 
 /// A helper function to merge one lineage into another, in DFS order.
 ///
```
Comment (richardliaw) [inline src/ray/raylet/lineage_cache.cc]: remove this helper?

File: src/ray/raylet/lineage_cache.cc [hunk 3]
```diff
@@ -180,34 +177,62 @@ void MergeLineageHelper(const TaskID &task_id, const Lineage &lineage_from,
...
+                                         std::unordered_set<TaskID> &subscribe_tasks) {
+  // If the entry is not found in the lineage to merge, then we stop since
+  // there is nothing to copy into the merged lineage.
+  auto entry = uncommitted_lineage.GetEntry(task_id);
+  if (!entry) {
+    return;
+  }
+  RAY_CHECK(entry->GetStatus() == GcsStatus::UNCOMMITTED_REMOTE);
+
+  // Insert a copy of the entry into lineage_to.
```
Comment (richardliaw) [inline src/ray/raylet/lineage_cache.cc]: update?

File: src/ray/raylet/lineage_cache.cc [hunk 4]
```diff
@@ -180,34 +177,62 @@ void MergeLineageHelper(const TaskID &task_id, const Lineage &lineage_from,
...
+  const auto &parent_ids = entry->GetParentTaskIds();
+  // If the insert is successful, then continue the DFS. The insert will fail
+  // if the new entry has an equal or lower GCS status than the current entry
+  // in lineage_to. This also prevents us from traversing the same node twice.
```
Comment (richardliaw) [inline src/ray/raylet/lineage_cache.cc]: update?

**Review by richardliaw — DISMISSED** [review]

**Comment by richardliaw** [comment]
Is there a test that specifically checks for the eviction invariant?

**Comment by stephanie-wang** [comment]
The `TestEvictUncommittedLineage` test case that I added does test the invariant, but we don't have a general way of verifying the invariant. That test case creates a chain of tasks and makes sure that the chain doesn't get evicted until all of the tasks have been committed. I'm thinking about adding another test case that has one task dependent on several other tasks, and making sure that the one task doesn't get evicted until all of its parents have been committed. I'm open to other ideas though :)

**Comment by stephanie-wang** [comment]
@richardliaw, I added another test for eviction safety (and also fixed the build).

**Review by richardliaw — APPROVED** [review]
changes lgtm, but tests errorring

**Comment by stephanie-wang** [comment]
Travis failures look unrelated.

**Pull request merged**